components/user-mgt/org.wso2.carbon.identity.user.pre.update.password.action/src/main/java/org/wso2/carbon/identity/user/pre/update/password/action/management/PreUpdatePasswordActionDTOModelResolver.java

```diff
@@ -0,0 +1,242 @@
+
+        try {
+            return getCertificateManagementService().addCertificate(certificate, tenantDomain);
+        } catch (CertificateMgtClientException e) {
+            throw new ActionDTOModelResolverClientException("Error while adding the certificate.", e.getDescription());
+        } catch (CertificateMgtException e) {
+            throw new ActionDTOModelResolverServerException("Error while adding the certificate.", e.getDescription(),
+                    e);
+        }
+    }
+
+    private Certificate getCertificate(String certificateId, String tenantDomain)
+            throws ActionDTOModelResolverException {
+
+        try {
+            return getCertificateManagementService().getCertificate(certificateId, tenantDomain);
+        } catch (CertificateMgtException e) {
+            throw new ActionDTOModelResolverServerException("Error while retrieving the certificate.",
+                    e.getDescription(), e);
+        }
+    }
+
+    private void resolveCertificateUpdate(ActionDTO updatingActionDTO, ActionDTO existingActionDTO,
+                                          Map<String, Object> properties, String tenantDomain)
+            throws ActionDTOModelResolverException {
+
+        Certificate updatingCertificate = (Certificate) updatingActionDTO.getProperty(CERTIFICATE);
+        Certificate existingCertificate = (Certificate) existingActionDTO.getProperty(CERTIFICATE);
+
+        if (updatingCertificate != null && updatingCertificate.getCertificateContent() != null) {
+            updateActionCertificate(updatingCertificate, existingCertificate, properties, tenantDomain);
+        } else if (existingCertificate != null && existingCertificate.getId() != null) {
+            // If a certificate is configured, and it is not updated; use the existing certificateId.
+            properties.put(CERTIFICATE, existingCertificate.getId());
+        }
+    }
+
+    private void updateActionCertificate(Certificate updatingCertificate, Certificate existingCertificate,
+                                         Map<String, Object> properties, String tenantDomain)
+            throws ActionDTOModelResolverException {
+
+        if (existingCertificate != null && existingCertificate.getId() != null) {
+            if (updatingCertificate.getCertificateContent().equals(StringUtils.EMPTY)) {
+                deleteCertificate(existingCertificate.getId(), tenantDomain);
+            } else {
+                updateCertificate(existingCertificate.getId(), updatingCertificate.getCertificateContent(),
+                        tenantDomain);
+
+                properties.put(CERTIFICATE, existingCertificate.getId());
+            }
+        } else {
+            Certificate certToBeAdded = buildCertificate(updatingCertificate.getId(), updatingCertificate);
+            properties.put(CERTIFICATE, addCertificate(certToBeAdded, tenantDomain));
+        }
+    }
```

Comment:
Looks to me this overall logic can be made more readable.
Let's do that over the next iteration an we need to prfioritize the delivery of the functionality

Reply:
Fixed